chemicaltoolbox/rdock/sort_filter.xml

```diff
@@ -1,38 +1,65 @@
-<tool id="rdock_sort_filter" name="SDF sort and filter" version="0.1.0">
+<tool id="rdock_sort_filter" name="SDF sort and filter" version="0.2.0">
     <description>using the sdsort provided with rDock</description>
     <macros>
         <import>rdock_macros.xml</import>
     </macros>
     <expand macro="requirements"/>
     <command><![CDATA[
 
-sdsort -n $descending -s -f'$sort_field' -id'$name_field' '$input' |
-  sdfilter -f'\$_COUNT <= $top' -s'$name_field' |
-  sdsort -n $descending -f'$sort_field' > '$output'
+cat '$input'
+#if $filters
+| sdfilter
+#for $i, $f in enumerate($filters)
+-f'$f.filter'
+#end for
+#end if
+#if $name_field
+| sdsort -n $descending -s -f'$sort_field' -id'$name_field'
+| sdfilter -f'\$_COUNT <= $top' -s'$name_field'
+#end if
+#if $global_sort and $sort_field
+| sdsort -n $descending -f'$sort_field'
+#end if
+> '$output'
 
     ]]></command>
 
     <inputs>
         <param type="data" name="input" format="sdf" label="Molecules" help="Molecules in SDF format"/>
-        <param name="top" type="integer" value="1" label="Number of records to keep in output" help="Number of best scoring records to keep"/>
-        <param name="sort_field" type="text" label="Field to sort on" optional="false" help="Name of the field to sort records by">
+        <repeat name="filters" title="Filters">
+            <param name="filter" type="text" label="Filter expression" optional="true" help="Perl expression for filter">
```

Comment:
The "Perl" here relevant? Its a simple math-expression, isn't it?
Do we need the repeat? Is that easier than just writing `or`?

Reply:
Technically it is a Perl expression. Any valid Perl code should be possible here. From the sdfilter help:
```
Filters SD records by data fields

Usage:	sdfilter -f'$<DataField> <Operator> <Value>' [-s<DataField>] [sdFiles]
or	sdfilter -f<filename> [-s<DataField>] [sdFiles]

Note:	Multiple filters are allowed and are OR'd together.
	Filters can be provided in a file, one per line.

	Standard Perl operators should be used. e.g.
	eq ne lt gt le ge for strings
	== != < > <= >= for numeric

	_REC (record #) is provided as a pseudo-data field
	if -s option is used, _COUNT (#occurrences of DataField) is provided as a pseudo-data field

	If SD file list not given, reads from standard input
	Output is to standard output
```

And yes, after I'd implemented the repeat I realised it wan't that useful.
A better approach might be to allow a multi-line text field and write that content to a config file. Again see the sdfilter help.

Reply:
@tdudgeon we need to make sure, __no__ arbitrary perl code can be executed here.
Maybe just allow those expressions (== != < > <= >=) all numbers and letters?

I don't mind the repeat, this is up to you.

Reply:
I removed the repeat from the filter as it didn't really add anything.

Reply:
> Maybe just allow those expressions (== != < > <= >=) all numbers and letters?

I'm not sure how best to restrict what can be entered as a filter. Certainly we must allow $ and text operators (eq ne lt gt le ge), probably round braces. Might also be nice to allow string matching (beginsWith, endsWith, contains - maybe using regexps?). But this might open things up too much.

Reply:
I'm hesitant to allow all those. In the end you can do a lot with perl and escape the commandline. I would rather restrict it and if you really have advanced needs like beginsWith, endsWith, contains the user should convert it into tabular and proceed from there.

What do you think?

Reply:
I'm happy with that. At least for starters. For now I just need to use simple expressions like '$SCORE  > 0.5'.
We can extend later if we need.